File: built-in plugins dsp/mb-wt-evolving.cpp
/**
 * MB Evolving Pad
 * Category : instrument
 * Type     : wavetable
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Slowly morphing wavetable pad
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_WT_EVOLVING_H
#define MB_WT_EVOLVING_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbWtEvolving : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-wt-evolving";
    static constexpr const char* PLUGIN_NAME    = "MB Evolving Pad";
    static constexpr const char* PLUGIN_TYPE    = "wavetable";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float position = 0f;  // range [0, 1]
    float lfoRate = 0.1f;  // range [0.01, 2]
    float volume = 0.7f;  // range [0, 1]
    };

    MbWtEvolving() = default;
    ~MbWtEvolving() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.position = std::clamp(params.position, 0f, 1f);
        params.lfoRate = std::clamp(params.lfoRate, 0.01f, 2f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Evolving Pad
        return input;
    }
};

#endif // MB_WT_EVOLVING_H

File: built-in plugins dsp/mb-wt-evolving.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbWtEvolvingPlugin: PluginDefinition = {
  id: "mb-wt-evolving",
  slug: "mb-wt-evolving",
  name: "MB Evolving Pad",
  category: "instrument",
  type: "wavetable",
  version: "1.0.0",
  description: "Slowly morphing wavetable pad",
  author: "Max Booster",
  grade: "A",
  oscillators: [{ type: "sine", detune: 0, gain: 0.6 }],
  envelope: { attack: 2.0, decay: 1.0, sustain: 0.8, release: 3.0 },
  parameters: [
    {
      id: "position",
      name: "Wave Position",
      type: "float",
      defaultValue: 0,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "lfoRate",
      name: "LFO Rate",
      type: "float",
      defaultValue: 0.1,
      minValue: 0.01,
      maxValue: 2,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { position: 0, lfoRate: 0.1, volume: 0.7 },
};

export default MbWtEvolvingPlugin;
